dbcache (phase2) enhancements

ptschip opened the pull request:
Only flush and delete pruned entries

    When reindexing or syncing the blockchain we want to keep
    any useful coins that may be dirty but have not been pruned yet.
    In this way sync and reindexing will be faster since we will
    not have to retrieve those coins again from leveldb.  Then rather 
    than flushing the entire in memory cache when it's full we trim it
    instead, just as we currently do when the chain is nearly synd.

ptschip committed:
Only flush and delete pruned entries

 When reindexing or syncing the blockchain we want to keep
 and useful coins that may be dirty but have not been pruned yet.
 In this way sync and reindexing will be faster since we will
 not have to retrieve those coins again from leveldb.

Don't process INV messages when reindexing

 There is no need to process any incoming INV's until reindex
 or import is complete.

When doing IBD or reindexing then trim the dbcache by 10%

 Trimming the dbcache by a larger amount during IBD reduces
 the frequency of flushing which was impactig performance
 negatively.

update max dbcache to 2GB and default dbcache to 500MB

 This allows a max dbcache on win32 systems of 2GB and bumps
 the dbache default to 500MB on all systems.

Flush the dbcache at least every 512MB increase

 To prevent excessively large flushes we flush the dbcache
 if we reach the limit or at least every 512MB increase
 since the last flush.

Do not delete coins when in BlocksOnly mode

 If we are running with fBlocksOnly set to true then we should not
 delete coins from the CoinsViewCache when we flush the to leveldb.
 We will need those coins for further validations just as we do when
 we are performing IBD because we are not receiving any coins into
 the cache from txns arriving in the mempool.

Only print to the log if nTrimmed > 0

 When printing out the number of trimmed entries from the dbcache
 only print to the log if there were entries that were trimmed. This
 keeps logging down.

Never Trim more than nMaxCacheIncreaseSinceLastFlush

 When a node operator does a reindex or IBD with a very large
 dbcache setting it may happen that the bytes trimmed ends up
 getting larger than the nMaxCacheIncreaseSinceLastFlush.  It
 doesn't makes sense to trim more than this amount so in such
 a case we set the bytes to be trimmed to be equal to
 nMaxCacheIncreaseSinceLastFlush.

sickpig commented:
failure on travis is due to formatting (`main.cpp`) issue not actual problems with the code

https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/264413219#L754

gandrewstone commented on src/main.cpp:
```diff
@@ -5996,7 +6009,7 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
 
 
-    else if (strCommand == NetMsgType::INV)
+    else if (strCommand == NetMsgType::INV && !fImporting && !fReindex)
```
the !fImporting and !fReindex here and in the other places need to be in a separate if statement below this one.  Otherwise the code will fall into the final "else" logic that is meant to be called only if the strCommand is unknown.

ptschip replied:
fixed.

gandrewstone commented on src/main.cpp:
```diff
@@ -6084,7 +6096,7 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
 
 
-    else if (strCommand == NetMsgType::GETDATA)
+    else if (strCommand == NetMsgType::GETDATA && !fImporting && !fReindex)
```
here too

ptschip replied:
fixed

gandrewstone commented on src/main.cpp:
```diff
@@ -6120,7 +6132,7 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
 
 
-    else if (strCommand == NetMsgType::GETBLOCKS)
+    else if (strCommand == NetMsgType::GETBLOCKS && !fImporting && !fReindex)
```
here too

ptschip replied:
fixed

ptschip committed:
Create smaller consistently size batches when flushing to leveldb

This prevents the large memory spikes associated with flushing
while at the same time preserving the good performance
characteritics of batching.  Essentially we are just creating
smaller batches of the same size rather than one unbounded
batch.

ptschip committed:
Do not process INV, GETDATA or GETBLOCKS when fReindex or fImporting

There is no need to process these messages when reindexing or
importing. We place the logic within the else statement so as
to not fall through to the final else logic if the strCommand
is known but we are importing or reindexing.

ptschip commented:
squashed
